app/jobs/download_ead_job.rb

```diff
@@ -36,18 +38,38 @@ def self.enqueue_one_by(aspace_repository_code:, updated_after: nil,
     enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:)
   end
 
-  # rubocop:disable Metrics/ParameterLists
-  def self.enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:)
-    resource_uris = AspaceClient.new.published_resource_uris(repository_id: aspace_repository_id, updated_after:)
+  # rubocop:disable Metrics/ParameterLists, Metrics/MethodLength
+  def self.enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:,
```

Comment:
@taylor-steve 100% take or leave but in order to get my head around this while reviewing I split out some of the logic like so -- no worries if you feel it does not add value.


```ruby
  # rubocop:disable Metrics/ParameterLists
  def self.enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:,
                   check_component_dates: false)
    resources = resources_to_enqueue(aspace_repository_id:, updated_after:, check_component_dates:)

    # debugger
    resources.each do |resource|
      enqueue_resource(aspace_repository_code:, ead_id: resource['ead_id'], resource_uri: resource['uri'],
                       data_dir:, index:, generate_pdf:)
    end
  end
  private_class_method :enqueue
  # rubocop:enable Metrics/ParameterLists

  def self.resources_to_enqueue(aspace_repository_id:, updated_after:, check_component_dates:)
    client = AspaceClient.new
    resource_uris = client.published_resource_uris(repository_id: aspace_repository_id, updated_after:)

    if check_component_dates
      uris_to_exclude = resource_uris.each { |resource| resource['uri'] }
      client.published_resource_with_updated_component_uris(
        repository_id: aspace_repository_id, updated_after: updated_after, uris_to_exclude: uris_to_exclude
      )
    else
      resource_uris
    end
  end
  private_class_method :resources_to_enqueue
```

Reply:
Thanks @marlo-longley. I'm wondering if I named `published_resource_with_updated_component_uris` poorly and/or the exclusion I added for performance is too confusing.

We need to enqueue *both* the result of `published_resource_uris` and `published_resource_with_updated_component_uris` because when using the exclusion param in this way, they contain mutually exclusive result sets. If we didn't use the exclusion, `published_resource_with_updated_component_uris` would return the complete set of what we want, but it could take hours to execute in many cases.

We could structure it more in the fashion you are here, but we'd have to return "resource sets" or something, so you'd have:
```ruby
def self.enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:, check_component_dates: false)
  resource_sets = resource_sets_to_enqueue(aspace_repository_id:, updated_after:, check_component_dates:)

  resource_sets.each do |resource_set|
    resource_set.each do |resource|
      enqueue_resource(aspace_repository_code:, ead_id: resource['ead_id'], resource_uri: resource['uri'],
                       data_dir:, index:, generate_pdf:)
    end
  end
end
```

Alternatively, we could flatten everything into one set, but then we'd lose the lazy evaluation of `AspaceQuery`.

I landed on enqueuing both sets in `enqueue` in this way because they are fully dependent on each other. If we have memory/perf to spare, we could ignore laziness and do something like this, which is slightly more readable:

```ruby
def self.enqueue(aspace_repository_id:, aspace_repository_code:, updated_after:, data_dir:, index:, generate_pdf:,
                 check_component_dates: false)
  aspace = AspaceClient.new
  resource_uris = aspace.published_resource_uris(repository_id: aspace_repository_id, updated_after:).each.to_a
  if check_component_dates
    resource_uris += aspace.published_resource_with_updated_component_uris(
      repository_id: aspace_repository_id, updated_after:, uris_to_exclude: resource_uris.map { |r| r['uri'] }
    ).each.to_a
  end

  resource_uris.each do |resource|
    enqueue_resource(aspace_repository_code:, ead_id: resource['ead_id'], resource_uri: resource['uri'],
                     data_dir:, index:, generate_pdf:)
  end
end
private_class_method :enqueue
```

If we're open to that idea, I'd probably consider pushing this back even further and have `AspaceQuery` return an array...